src/realm/db.cpp

```diff
@@ -78,304 +78,192 @@ namespace {
+    // is the only thread/process that has access to the VersionList. It is currently
+    // called from init_versioning(), which is called by DB::open() under the
+    // condition that it is the session initiator and under guard by the control mutex,
+    // thus ensuring the precondition.
+    // It is most likely not suited for any other use.
+    ReadCount& init_versioning() noexcept
     {
-        REALM_ASSERT(!is_full());
-        return data[next()];
+        int num_entries = entries;
+        entries = 0;
+        newest = oldest = freelist = nil;
+        expand_to(num_entries);
+        ReadCount& r = allocate_entry();
+        r.count_live = r.count_frozen = 0;
+        return r;
     }
 
-    void use_next() noexcept
+    bool is_full() const noexcept
     {
-        atomic_dec(get_next().count); // .store_release(0);
-        put_pos.store(uint32_t(next()), std::memory_order_release);
+        return freelist == nil;
     }
 
-    void cleanup() noexcept
+    void purge_versions(uint64_t& oldest_v, uint64_t& oldest_live_v, TopRefMap& top_refs,
+                        int& num_unreachable) noexcept
     {
-        // invariant: entry held by put_pos has count > 1.
-        // std::cout << "cleanup: from " << old_pos << " to " << put_pos.load_relaxed();
-        // dump();
-        while (old_pos.load(std::memory_order_relaxed) != put_pos.load(std::memory_order_relaxed)) {
-            const ReadCount& r = get(old_pos.load(std::memory_order_relaxed));
-            if (!atomic_one_if_zero(r.count))
-                break;
-            auto next_ndx = get(old_pos.load(std::memory_order_relaxed)).next;
-            old_pos.store(next_ndx, std::memory_order_relaxed);
+        // run through list of versions and reclaim unused entries.
+        // buld a map of reachable versions on the way.
+        // there is always at least one live entry
+        auto it = oldest;
+        bool looking_for_oldest_live = true;
+        bool have_seen_a_reachable_version = false;
+        REALM_ASSERT(oldest != nil);
+        while (it != nil) {
+            auto& r = get(it);
+            it = r.newer;
+            if (looking_for_oldest_live && r.count_live != 0) {
+                looking_for_oldest_live = false;
+                oldest_v = get(oldest).version;
+                oldest_live_v = r.version;
+                REALM_ASSERT(oldest_v <= oldest_live_v);
+            }
+            if (r.count_frozen == 0 && r.count_live == 0) {
+                // only count in-between versions... this results in fewer runs
+                // of the backdating algorithm in group_writer.
+                if (have_seen_a_reachable_version)
+                    num_unreachable++;
+                r.version = 0;
+                free_entry(index_of(r));
+            }
+            else {
+                have_seen_a_reachable_version = true;
+                top_refs.emplace(r.version, VersionInfo{to_ref(r.current_top), to_ref(r.filesize)});
```

Comment:
Might be worth using `emplace_hint(top_refs.end(), ...)` to make this O(n) rather than O(n log n). Although some other suggestions invalidate this one.